build_tools/benchmark_logs.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
+#  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+#  This source code is licensed under both the GPLv2 (found in the
+#  COPYING file in the root directory) and Apache 2.0 License
+#  (found in the LICENSE.Apache file in the root directory).
+
+'''Take benchmark logs and output them as a Multipart MIME:
+    - Designed to bypass failed uploads to artifacts
+    - Might just do what we need anyway
+'''
+
+import mimetypes
+import os
+import sys
+import argparse
+from email.message import EmailMessage
+
+
+class BenchmarkLogger(object):
+
+    def __init__(self, args):
+        self.logdirectory = args.logdirectory
+
+    def build_message(self):
+        message = EmailMessage()
```

Comment:
do we have to use MIME data? does grafana support rest api?
I'm trying to setup an AWS grafana service, but seems it only support SAML or SSO auth (I don't have permission to create SSO, for SAML, not sure how we could setup that): https://g-b7e3136b85.grafana-workspace.us-west-2.amazonaws.com/login
Another option is use Opensearch as a service, I created a test one: shorturl.at/mrHLR username: rocksdb pass: 73l%X0Vjxzv6 the endpoint is public available, which is only for test. (for prod, we can use vpc for security.)

Reply:
We are sending data to Graphite (the time series DB back end attached to Grafana) via pickle (Python data encoding) on a standard port on the Graphite service. See https://graphiteapp.org/
The MIME-ification code dumps a multipart MIME message of the output files in the logs; this was needed as a workaround because the AWS box would not play nice with uploading CircleCI artefacts; I had initially understood that we would access the results via the Circle API, so there's some code to read those back via the Circle API and unpack the MIME - it's not used in the current path (because we push results to Graphite as end of the benchmark) but I wanted to keep it in case it was needed.

The Grafana (and Graphite) service the demo uses is run using docker and is in this repo https://github.com/alanpaxton/grafanagraphite so you can simply set up a VM, follow the instructions in the comments in docker_compose.yml, and the service will be available to push benchmark results into; Grafana dashboard config in the GUI is then very pointy-clicky. Does that help ?

Reply:
We don't want to host and manage a grafana service (or similar metrics service) if a public service is available, either AWS or GCP is good (assuming the benchmark will run on the same platform).

Reply:
@jay-zhuang At the moment we are hosting a Graphite+Graphana Docker Container as a temporary solution on our infrastructure here at Evolved Binary. The goal will be to move this to AWS when you are ready. To do so we will likely need you to provision:
1. a new EC2 instance for us (as this should not be run on the same host that is performing the benchmarks).
2. an S3 bucket where we can store the Graphite database backups.

I think for the time being to prove that this works the temporary solution is okay for us to merge this PR.

Once you provide us with the access credentials for the additional EC2 instance and S3 bucket we can get that setup, and we can then send a subsequent PR which changes the URL of where to send the data to Graphite.

Reply:
Here is the host name: `ubuntu@ec2-34-223-42-134.us-west-2.compute.amazonaws.com`
I added both your and @alanpaxton 's public keys to it, please let me know if it work.

For S3, I don't know how to setup the security for backup and that's also one of the reasons that we don't want to run the DB by ourself.

Anyway I think it's good start, thanks for your effort. I'm going to accept the PR to start the test. But my concerns are not all addressed:
1. why not simpler API like rest which can be send with curl
2. why circleci package is needed in the script to send the data, which is to get output from the previous step, that could be by getting it from the local host?
3. use existing SaaS service instead of running our own.

Reply:
@jay-zhuang if you don't want us to run this in EC2... where would you like it run? I am not aware of a Grafana/Graphite SaSS service for this that the Rocks project has purchased - do you know otherwise? If so can you send Alan and I the access credentials for that service?

Reply:
My comment above:
> I'm trying to setup an AWS grafana service, but seems it only support SAML or SSO auth (I don't have permission to create SSO, for SAML, not sure how we could setup that): [g-b7e3136b85.grafana-workspace.us-west-2.amazonaws.com/login](https://g-b7e3136b85.grafana-workspace.us-west-2.amazonaws.com/login)
Another option is use Opensearch as a service, I created a test one: shorturl.at/mrHLR username: rocksdb pass: 73l%X0Vjxzv6 the endpoint is public available, which is only for test. (for prod, we can use vpc for security.)

Opensearch is aws version elasticsearch, AWS still provide 7.x elasticsearch service with kibana, which should also fit our usecase. (grafana is just a fork of kibana).

Reply:
@jay-zhuang Sorry I missed some of the details in that comment above.
May I politely suggest that it may be better to send us username/password/url directly to me rather than in this public channel.

At the moment, as previously requested, we have designed for Grafana and Graphite.
Whilst AWS provide a managed Grafana service, you have to separately setup and configure a database (datasource in Grafana parlance) to receive and hold the data that Grafana visualises. Unfortunately AWS do not provide Graphite as a managed service, the closest thing they have as far as I can see is Amazon Timestream.

AWS Timestream can be used as a datasource for Grafana, however you still have to configure it for your data retention and backup policy.

It is not at all clear to me whether you can use HTTP to directly write data into Timestream, their documentation seems to indicate that you should use one of their language specific SDK's (presumably Python for us) - https://docs.aws.amazon.com/timestream/latest/developerguide/API_WriteRecords.html

Regardless, I think like most AWS services, the more you use the more you need, so for Grafana, it does look like you will have to setup and integrated with SSO or SAML (AWS also have services for these!).

It also have doubts whether with the AWS managed Grafana service, you can actually make your dashboards public or not (whereas it seems you can if you self-host the OSS edition), see: 
1. https://community.grafana.com/t/create-public-grafana-dashboards-without-login/34458/7
2. https://grafana.com/docs/grafana/latest/sharing/share-dashboard/

Regards Elastic/OpenSearch - I am not sure it is the right tool for this, as Lucene (the underlying db) is not really designed for Time Series data, hence us selecting the Graphite/Grafana pairing.

Reply:
I think Elastic is a good fit. Here is a simple script to run the benchmark and push the data to es

```python
#!/usr/bin/env python3
import json
import os
import pandas
import pprint
import requests
import subprocess


def run_bench():
  cmd = './tools/benchmark.sh fillseq_enable_wal'
  subprocess.run(cmd, env={'DB_DIR': '/tmp/rocksdb-benchmark-datadir',
                           'WAL_DIR': '/tmp/rocksdb-benchmark-waldir',
                           'OUTPUT_DIR': '/tmp/benchmark-results',
                           'NUM_KEYS': '1000000',
                           'LD_LIBRARY_PATH': '/usr/local/lib'}, shell=True,
                 check=True, capture_output=True, text=True)
  data = pandas.read_csv('/tmp/benchmark-results/report.tsv', sep='\t',
                         header=0)
  data.rename(mapper=lambda x: x.replace('.', '_'), axis='columns',
              inplace=True)
  d1 = data.iloc[-1].to_json()
  parsed = json.loads(d1)
  pprint.pprint(parsed)
  response = requests.post(
    'https://search-rocksdb-bench-k2izhptfeap2hjfxteolsgsynm.us-west-2.es.amazonaws.com/bench_test3/_doc',
    json=parsed, auth=(os.environ['ES_USER'], os.environ['ES_PASS']))
  print(response.status_code)
  print(response.text)


if __name__ == '__main__':
  for i in range(100000):
    print('=' * 40, i)
    run_bench()
```

here is the dashboard:
https://search-rocksdb-bench-k2izhptfeap2hjfxteolsgsynm.us-west-2.es.amazonaws.com/_dashboards/goto/587200a751b0740a371110c464b4c135?security_tenant=global

<img width="2651" alt="image" src="https://user-images.githubusercontent.com/33968/163737431-836bd4f9-a6ea-4bb9-ac6e-325c2f78006d.png">

Reply:
@alanpaxton Done!

Reply:
@adamretter OpenSearch-upload-only version just pushed. Hopefully converging on something that works now.